// azure-cli/lib/commands/arm/keyvault/keyvault.key.js
/*** Generated by streamline 0.10.17 (callbacks) - DO NOT EDIT ***/ "use strict"; var __rt=require('streamline/lib/callbacks/runtime').runtime(__filename, false),__func=__rt.__func,__cb=__rt.__cb,__tryCatch=__rt.__tryCatch; var __ = require("underscore");




















var util = require("util");
var fs = require("fs");
var forge = require("node-forge");

var profile = require("../../../util/profile");
var utils = require("../../../util/utils");
var kvUtils = require("./kv-utils");

var $ = utils.getLocaleString;

var KEY_DEST_TYPE_MAP = {
  Software: "RSA",
  HSM: "RSA-HSM"};

var KEY_DESTS = Object.keys(KEY_DEST_TYPE_MAP);

var KEY_OPS = ["encrypt","decrypt","sign","verify","wrapKey","unwrapKey",];

exports.init = function(cli) {
  var log = cli.output;

  var key = cli.category("keyvault").category("key").description($("Commands to manage keys in the Azure Key Vault service"));


  key.command("list [vault-name]").description($("Lists keys of a vault")).usage("[options] <vault-name>").option("-u, --vault-name <vault-name>", $("the vault name")).execute(function __1(vaultName, options, _) { var client, keys, progress, result; var __frame = { name: "__1", line: 50 }; return __func(_, this, arguments, __1, 2, __frame, function __$__1() {









      log.verbose(("arguments: " + JSON.stringify({
        vaultName: vaultName,
        options: options })));


      options.vaultName = (options.vaultName || vaultName);

      if (!options.vaultName) {
        return _(null, cli.missingArgument("vault-name")); } ;






      options.vaultUri = createVaultUri(options);
      client = createClient(options);

      keys = [];
      progress = cli.interaction.progress(util.format($("Loading keys of vault %s"), options.vaultUri)); return (function ___(__then) { (function ___(_) { __tryCatch(_, function __$__1() {

            return client.getKeys(options.vaultUri, null, __cb(_, __frame, 27, 28, function ___(__0, __1) { result = __1; return (function __$__1(__then) {
                if (result) {
                  keys = result; return (function ___(__break) { var __more; var __loop = __cb(_, __frame, 0, 0, function __$__1() { __more = false;

                      var __4 = (result && result.nextLink); if (__4) {
                        log.verbose(util.format($("Found %d keys, loading more"), keys.length));
                        return client.getKeysNext(result.nextLink, __cb(_, __frame, 33, 28, function ___(__0, __2) { result = __2;
                          if (result) {
                            keys = keys.concat(result); } ; while (__more) { __loop(); }; __more = true; }, true)); } else { __break(); } ; }); do { __loop(); } while (__more); __more = true; })(__then); } else { __then(); } ; })(function __$__1() { _(null, null, true); }); }, true)); }); })(function ___(__e, __r, __cont) { (function ___(__then) { __tryCatch(_, function __$__1() {




              progress.end(); __then(); }); })(function ___() { __tryCatch(_, function ___() { if (__cont) { __then(); } else { _(__e, __r); }; }); }); }); })(function ___() { __tryCatch(_, function __$__1() {


          log.table(keys, showKeyRow);
          log.info(util.format($("Found %d keys"), keys.length)); _(); }); }); }); });


  key.command("list-versions [vault-name] [key-name]").description($("Lists key versions")).usage("[options] <vault-name> [key-name>]").option("-u, --vault-name <vault-name>", $("the vault name")).option("-k, --key-name <key-name>", $("lists only versions of this key")).execute(function __2(vaultName, keyName, options, _) { var client, keys, keyIdentifier, keyVersions, progress, result, i, j; var __frame = { name: "__2", line: 102 }; return __func(_, this, arguments, __2, 3, __frame, function __$__2() {










      log.verbose(("arguments: " + JSON.stringify({
        vaultName: vaultName,
        keyName: keyName,
        options: options })));


      options.vaultName = (options.vaultName || vaultName);
      options.keyName = (options.keyName || keyName);

      if (!options.vaultName) {
        return _(null, cli.missingArgument("vault-name")); } ;






      options.vaultUri = createVaultUri(options);
      client = createClient(options); return (function __$__2(__then) {



        if (!options.keyName) {
          keys = [];
          progress = cli.interaction.progress(util.format($("Loading keys of vault %s"), options.vaultUri)); return (function ___(__then) { (function ___(_) { __tryCatch(_, function __$__2() {

                return client.getKeys(options.vaultUri, null, __cb(_, __frame, 32, 30, function ___(__0, __1) { result = __1; return (function __$__2(__then) {
                    if ((result && result.length)) {
                      i = 0; var __8 = false; return (function ___(__break) { var __more; var __loop = __cb(_, __frame, 0, 0, function __$__2() { __more = false; if (__8) { ++i; } else { __8 = true; } ; var __7 = (i < result.length); if (__7) {
                            keyIdentifier = kvUtils.parseKeyIdentifier(result[i].kid);
                            return getKeyVersions(client, keyIdentifier.vaultUri, keyIdentifier.name, __cb(_, __frame, 36, 28, function ___(__0, __2) { keyVersions = __2;
                              keys = keys.concat(keyVersions); while (__more) { __loop(); }; __more = true; }, true)); } else { __break(); } ; }); do { __loop(); } while (__more); __more = true; })(function __$__2() { return (function ___(__break) { var __more; var __loop = __cb(_, __frame, 0, 0, function __$__2() { __more = false;


                            var __10 = (result && result.nextLink); if (__10) {
                              log.verbose(util.format($("Found %d keys, loading more"), keys.length));
                              return client.getKeysNext(result.nextLink, __cb(_, __frame, 42, 30, function ___(__0, __3) { result = __3; return (function __$__2(__then) {
                                  if ((result && result.length)) {
                                    j = 0; var __13 = false; return (function ___(__break) { var __more; var __loop = __cb(_, __frame, 0, 0, function __$__2() { __more = false; if (__13) { ++j; } else { __13 = true; } ; var __12 = (j < result.length); if (__12) {
                                          keyIdentifier = kvUtils.parseKeyIdentifier(result[j].kid);
                                          return getKeyVersions(client, keyIdentifier.vaultUri, keyIdentifier.name, __cb(_, __frame, 46, 32, function ___(__0, __4) { keyVersions = __4;
                                            keys = keys.concat(keyVersions); while (__more) { __loop(); }; __more = true; }, true)); } else { __break(); } ; }); do { __loop(); } while (__more); __more = true; })(__then); } else { __then(); } ; })(function __$__2() { while (__more) { __loop(); }; __more = true; }); }, true)); } else { __break(); } ; }); do { __loop(); } while (__more); __more = true; })(__then); }); } else { __then(); } ; })(function __$__2() { _(null, null, true); }); }, true)); }); })(function ___(__e, __r, __cont) { (function ___(__then) { __tryCatch(_, function __$__2() {





                  progress.end(); __then(); }); })(function ___() { __tryCatch(_, function ___() { if (__cont) { __then(); } else { _(__e, __r); }; }); }); }); })(function ___() { __tryCatch(_, __then); }); } else {


          progress = cli.interaction.progress(util.format($("Loading keys of vault %s"), options.vaultUri)); return (function ___(__then) { (function ___(_) { __tryCatch(_, function __$__2() {

                return getKeyVersions(client, options.vaultUri, options.keyName, __cb(_, __frame, 58, 17, function ___(__0, __5) { keys = __5; _(null, null, true); }, true)); }); })(function ___(__e, __r, __cont) { (function ___(__then) { __tryCatch(_, function __$__2() {

                  progress.end(); __then(); }); })(function ___() { __tryCatch(_, function ___() { if (__cont) { __then(); } else { _(__e, __r); }; }); }); }); })(function ___() { __tryCatch(_, __then); }); } ; })(function __$__2() {



        log.table(keys, showKeyRow);
        log.info(util.format($("Found %d keys"), keys.length)); _(); }); }); });


  key.command("create [vault-name] [key-name]").description($("Creates a key in the vault")).usage("[options] <vault-name> <key-name>").option("-u, --vault-name <vault-name>", $("the vault name")).option("-k, --key-name <key-name>", $("name of the key to be created; if already exists, a new key version is generated")).option("-d, --destination <destination>", util.format($("tells if the key is software-protected or HSM-protected; valid values: [%s]"), KEY_DESTS.join(", "))).option("--enabled <boolean>", $("tells if the key should be enabled; valid values: [false, true]; default is true")).option("-e, --expires <datetime>", $("key expiration time, expressed in RFC-1123/ISO8601 date format")).option("-n, --not-before <datetime>", $("time before which key cannot be used, expressed in RFC-1123/ISO8601 date format")).option("-o, --key-ops <key-ops>", util.format($("JSON-encoded array of strings representing key operations; each string can be one of [%s]"), KEY_OPS.join(", "))).option("-t, --tags <tags>", $("Tags to set on the key. Can be multiple in the format 'name=value'. Name is required and value is optional. For example, -t tag1=value1;tag2")).execute(function __3(vaultName, keyName, options, _) { var keyVersion, client, requestOptions, key, keyIdentifier, progress; var __frame = { name: "__3", line: 181 }; return __func(_, this, arguments, __3, 3, __frame, function __$__3() {

















      parseKeyPropertiesArguments(vaultName, keyName, keyVersion, options, true);
      options.kty = (options.hsm ? "RSA-HSM" : "RSA");





      options.vaultUri = createVaultUri(options);
      client = createClient(options);

      requestOptions = {
        keyOps: options.keyOps,
        keyAttributes: {
          enabled: options.enabled,
          notBefore: options.notBefore,
          expires: options.expires },

        tags: options.tags };


      log.verbose(("request options: " + JSON.stringify(requestOptions)));


      keyIdentifier = getKeyIdentifier(options);
      progress = cli.interaction.progress(util.format($("Creating key %s"), keyIdentifier)); return (function ___(__then) { (function ___(_) { __tryCatch(_, function __$__3() {

            return client.createKey(options.vaultUri, options.keyName, options.kty, requestOptions, __cb(_, __frame, 33, 21, function ___(__0, __1) { key = __1; _(null, null, true); }, true)); }); })(function ___(__e, __r, __cont) { (function ___(__then) { __tryCatch(_, function __$__3() {

              progress.end(); __then(); }); })(function ___() { __tryCatch(_, function ___() { if (__cont) { __then(); } else { _(__e, __r); }; }); }); }); })(function ___() { __tryCatch(_, function __$__3() {


          showKey(key); _(); }); }); }); });


  key.command("import [vault-name] [key-name]").description($("Imports an existing key into a vault")).usage("[options] <vault-name> <key-name>").option("-u, --vault-name <vault-name>", $("the vault name")).option("-k, --key-name <key-name>", $("name of the key to be imported; if already exists, a new key version is generated")).option("--pem-file <file-name>", $("name of a PEM file containing the key to be imported; the file must not be password protected")).option("--byok-file <file-name>", $("name of a BYOK file containing the key to be imported")).option("-p, --password <password>", $("password of key file; if not informed and the file is encrypted, will prompt")).option("-d, --destination <destination>", util.format($("tells if the key is software-protected or HSM-protected; valid values: [%s]"), KEY_DESTS.join(", "))).option("--enabled <boolean>", $("tells if the key should be enabled; valid values: [false, true]; default is true")).option("-e, --expires <datetime>", $("key expiration time, expressed in RFC-1123/ISO8601 date format")).option("-n, --not-before <datetime>", $("time before which key cannot be used, expressed in RFC-1123/ISO8601 date format")).option("-o, --key-ops <key-ops>", util.format($("JSON-encoded array of strings representing key operations; each string can be one of [%s]"), KEY_OPS.join(", "))).option("-t, --tags <tags>", $("Tags to set on the key. Can be multiple in the format 'name=value'. Name is required and value is optional. For example, -t tag1=value1;tag2")).execute(function __4(vaultName, keyName, options, _) { var keyVersion, v, requestKey, requestOptions, data, keyFile, keyInfo, encrypted, pwdMsg, password, client, key, progress; var __frame = { name: "__4", line: 236 }; return __func(_, this, arguments, __4, 3, __frame, function __$__4() {




















      parseKeyPropertiesArguments(vaultName, keyName, keyVersion, options, options.pemFile);

      v = 0;
      if (options.pemFile) { v++; } ;
      if (options.byokFile) { v++; } ;

      if ((v != 1)) {
        v = ["zero","one","two",][v];
        log.error(util.format($("Expecting exactly one of the following, but %s were informed:"), v));
        log.error($("    --pem-file <file-name>"));
        log.error($("    --byok-file <file-name>"));
        return _(new Error($("Could not establish key to import from command arguments"))); } ;


      if (((options.byokFile && options.destination) && !options.hsm)) {
        return _(new Error(util.format($("Value of parameter --destination (%s) is incompatible with input key type (BYOK)."), options.destination))); } ;






      requestKey = {
        keyOps: options.keyOps };


      requestOptions = {
        hsm: options.hsm,
        keyAttributes: {
          enabled: options.enabled,
          notBefore: options.notBefore,
          expires: options.expires },

        tags: options.tags }; return (function __$__4(__then) {




        if (options.pemFile) {

          requestKey.kty = "RSA";
          log.verbose(("reading " + options.pemFile));
          data = fs.readFileSync(options.pemFile);

          encrypted = isPemEncrypted(data); return (function __$__4(__then) {
            if (encrypted) {
              pwdMsg = util.format($("Password for %s: "), options.pemFile);
              return cli.interaction.promptPasswordOnceIfNotGiven(pwdMsg, options.password, __cb(_, __frame, 54, 41, function ___(__0, __1) { password = __1;
                keyInfo = forge.pki.decryptRsaPrivateKey(data, password); __then(); }, true)); } else {

              keyInfo = forge.pki.privateKeyFromPem(data);
              if (options.password) {
                log.warn(util.format($("File %s is not password protected, the --password flag is extraneous and was ignored"), options.pemFile)); } ; __then(); } ; })(function __$__4() {



            log.verbose("setting RSA parameters from PEM data");
            setRsaParameters(requestKey, keyInfo);
            keyFile = options.pemFile; __then(); }); } else {



          requestKey.kty = "RSA-HSM";
          log.verbose(("reading " + options.byokFile));
          log.verbose("setting BYOK parameters from file");
          requestKey.t = fs.readFileSync(options.byokFile);
          keyFile = options.byokFile;

          if (options.password) {
            log.warn($("The --password flag is extraneous and was ignored")); } ; __then(); } ; })(function __$__4() {








        options.vaultUri = createVaultUri(options);
        client = createClient(options);

        log.verbose(("request options: " + JSON.stringify(requestOptions, null, " ")));


        progress = cli.interaction.progress(util.format($("Importing file %s into key %s"), keyFile, options.keyName)); return (function ___(__then) { (function ___(_) { __tryCatch(_, function __$__4() {

              return client.importKey(options.vaultUri, options.keyName, requestKey, requestOptions, __cb(_, __frame, 93, 21, function ___(__0, __2) { key = __2; _(null, null, true); }, true)); }); })(function ___(__e, __r, __cont) { (function ___(__then) { __tryCatch(_, function __$__4() {

                progress.end(); __then(); }); })(function ___() { __tryCatch(_, function ___() { if (__cont) { __then(); } else { _(__e, __r); }; }); }); }); })(function ___() { __tryCatch(_, function __$__4() {


            showKey(key); _(); }); }); }); }); });


  key.command("set-attributes [vault-name] [key-name] [key-version]").description($("Changes attributes of an existing key")).usage("[options] <vault-name> <key-name> [key-version]").option("-u, --vault-name <vault-name>", $("the vault name")).option("-k, --key-name <key-name>", $("name of the key to be modified")).option("-r, --key-version <key-version>", $("the version to be modified; if omited, modifies only the most recent")).option("--enabled <boolean>", $("if informed, command will change the enabled state; valid values: [false, true]")).option("-e, --expires <datetime>", $("if informed, command will change secret expiration time; expressed in RFC-1123/ISO8601 date format, or null to clear the value")).option("-n, --not-before <datetime>", $("if informed, command will change time before which secret cannot be used; expressed in RFC-1123/ISO8601 date format, or null to clear the value")).option("-o, --key-ops <key-ops>", util.format($("if informed, command will change valid operations; must be JSON-encoded array of strings representing key operations; each string can be one of [%s]"), KEY_OPS.join(", "))).option("-t, --tags <tags>", $("Tags to set on the key. Can be multiple in the format 'name=value'. Name is required and value is optional. For example, -t tag1=value1;tag2")).option("--reset-tags", $("remove previously existing tags; can combined with --tags")).execute(function __5(vaultName, keyName, keyVersion, options, _) { var informed, client, key, keyIdentifier, currentTags, request, progress; var __frame = { name: "__5", line: 349 }; return __func(_, this, arguments, __5, 4, __frame, function __$__5() {


















      informed = {
        enabled: (options.enabled || false),
        expires: (options.expires || false),
        notBefore: (options.notBefore || false),
        keyOps: (options.keyOps || false),
        tags: (options.tags || false),
        resetTags: (options.resetTags || false) };


      parseKeyPropertiesArguments(vaultName, keyName, keyVersion, options, false);





      options.vaultUri = createVaultUri(options);
      client = createClient(options);


      keyIdentifier = getKeyIdentifier(options); return (function __$__5(__then) {

        if (informed.tags) { return (function __$__5(__then) {



            if (!informed.resetTags) {


              log.info(util.format($("Getting key %s"), keyIdentifier));
              return client.getKey(keyIdentifier, __cb(_, __frame, 36, 23, function ___(__0, __1) { key = __1;
                currentTags = key.tags;
                if (!currentTags) {

                  currentTags = { }; } ;

                options.tags = kvUtils.mergeTags(currentTags, options.tags); __then(); }, true)); } else { __then(); } ; })(__then); } else {







          if (informed.resetTags) {


            informed.tags = true;
            options.tags = { }; } ; __then(); } ; })(function __$__5() {









        request = {
          keyAttributes: { } };


        if (informed.keyOps) { request.keyOps = options.keyOps; } ;
        if (informed.enabled) { request.keyAttributes.enabled = options.enabled; } ;
        if (informed.notBefore) { request.keyAttributes.notBefore = options.notBefore; } ;
        if (informed.expires) { request.keyAttributes.expires = options.expires; } ;
        if (informed.tags) { request.tags = options.tags; } ;





        log.verbose(("request: " + JSON.stringify(request, null, " ")));

        progress = cli.interaction.progress(util.format($("Updating key %s"), keyIdentifier)); return (function ___(__then) { (function ___(_) { __tryCatch(_, function __$__5() {

              return client.updateKey(keyIdentifier, request, __cb(_, __frame, 82, 21, function ___(__0, __2) { key = __2; _(null, null, true); }, true)); }); })(function ___(__e, __r, __cont) { (function ___(__then) { __tryCatch(_, function __$__5() {

                progress.end(); __then(); }); })(function ___() { __tryCatch(_, function ___() { if (__cont) { __then(); } else { _(__e, __r); }; }); }); }); })(function ___() { __tryCatch(_, function __$__5() {


            showKey(key); _(); }); }); }); }); });


  key.command("show [vault-name] [key-name] [key-version]").description($("Shows properties of a vault key")).usage("[options] <vault-name> <key-name> [key-version]").option("-u, --vault-name <vault-name>", $("the vault name")).option("-k, --key-name <key-name>", $("the key name")).option("-r, --key-version <key-version>", $("the key version; if omited, uses the most recent")).execute(function __6(vaultName, keyName, keyVersion, options, _) { var client, keyIdentifier, progress; var __frame = { name: "__6", line: 445 }; return __func(_, this, arguments, __6, 4, __frame, function __$__6() {











      log.verbose(("arguments: " + JSON.stringify({
        vaultName: vaultName,
        keyName: keyName,
        keyVersion: keyVersion,
        options: options })));


      options.vaultName = (options.vaultName || vaultName);
      options.keyName = (options.keyName || keyName);
      options.keyVersion = (options.keyVersion || keyVersion);

      if (!options.vaultName) {
        return _(null, cli.missingArgument("vault-name")); } ;


      if (!options.keyName) {
        return _(null, cli.missingArgument("key-name")); } ;






      options.vaultUri = createVaultUri(options);
      client = createClient(options);

      keyIdentifier = getKeyIdentifier(options);
      progress = cli.interaction.progress(util.format($("Getting key %s"), keyIdentifier)); return (function ___(__then) { (function ___(_) { __tryCatch(_, function __$__6() {

            return client.getKey(keyIdentifier, __cb(_, __frame, 35, 21, function ___(__0, __1) { key = __1; _(null, null, true); }, true)); }); })(function ___(__e, __r, __cont) { (function ___(__then) { __tryCatch(_, function __$__6() {

              progress.end(); __then(); }); })(function ___() { __tryCatch(_, function ___() { if (__cont) { __then(); } else { _(__e, __r); }; }); }); }); })(function ___() { __tryCatch(_, function __$__6() {


          showKey(key); _(); }); }); }); });


  key.command("delete [vault-name] [key-name]").description($("Deletes a key from the vault")).usage("[options] <vault-name> <key-name>").option("-u, --vault-name <vault-name>", $("the vault name")).option("-k, --key-name <key-name>", $("the key name")).option("-q, --quiet", $("quiet mode (do not ask for delete confirmation)")).execute(function __7(vaultName, keyName, options, _) { var client, key, keyIdentifier, progress; var __frame = { name: "__7", line: 494 }; return __func(_, this, arguments, __7, 3, __frame, function __$__7() {











      log.verbose(("arguments: " + JSON.stringify({
        vaultName: vaultName,
        keyName: keyName,
        options: options })));


      options.vaultName = (options.vaultName || vaultName);
      options.keyName = (options.keyName || keyName);

      if (!options.vaultName) {
        return _(null, cli.missingArgument("vault-name")); } ;


      if (!options.keyName) {
        return _(null, cli.missingArgument("key-name")); } ; return (function __$__7(_) {


        var __1 = !options.quiet; if (!__1) { return _(null, __1); } ; return cli.interaction.confirm(util.format($("Delete key %s from vault %s? [y/n] "), options.keyName, options.vaultName), __cb(_, __frame, 23, 45, function ___(__0, __3) { var __2 = !__3; return _(null, __2); }, true)); })(__cb(_, __frame, -493, 17, function ___(__0, __2) { return (function __$__7(__then) { if (__2) {
            return _(new Error($("Aborted by user"))); } else { __then(); } ; })(function __$__7() {






          options.vaultUri = createVaultUri(options);
          client = createClient(options);


          keyIdentifier = getKeyIdentifier(options);
          progress = cli.interaction.progress(util.format($("Deleting key %s"), keyIdentifier)); return (function ___(__then) { (function ___(_) { __tryCatch(_, function __$__7() {

                return client.deleteKey(options.vaultUri, options.keyName, __cb(_, __frame, 38, 21, function ___(__0, __3) { key = __3; _(null, null, true); }, true)); }); })(function ___(__e, __r, __cont) { (function ___(__then) { __tryCatch(_, function __$__7() {

                  progress.end(); __then(); }); })(function ___() { __tryCatch(_, function ___() { if (__cont) { __then(); } else { _(__e, __r); }; }); }); }); })(function ___() { __tryCatch(_, function __$__7() {


              showKey(key); _(); }); }); }); }, true)); }); });


  key.command("backup [vault-name] [key-name] [output-file]").description($("Generates a protected backup of a vault key")).usage("[options] <vault-name> <key-name> <output-file>").option("-u, --vault-name <vault-name>", $("the vault name")).option("-k, --key-name <key-name>", $("the key name")).option("-f, --output-file <output-file>", $("name of the binary file that will contain backup data")).execute(function __8(vaultName, keyName, outputFile, options, _) { var client, blob, keyIdentifier, progress; var __frame = { name: "__8", line: 546 }; return __func(_, this, arguments, __8, 4, __frame, function __$__8() {











      log.verbose(("arguments: " + JSON.stringify({
        vaultName: vaultName,
        keyName: keyName,
        outputFile: outputFile,
        options: options })));


      options.vaultName = (options.vaultName || vaultName);
      options.keyName = (options.keyName || keyName);
      options.outputFile = (options.outputFile || outputFile);

      if (!options.vaultName) {
        return _(null, cli.missingArgument("vault-name")); } ;


      if (!options.keyName) {
        return _(null, cli.missingArgument("key-name")); } ;


      if (!options.outputFile) {
        return _(null, cli.missingArgument("output-file")); } ;






      options.vaultUri = createVaultUri(options);
      client = createClient(options);


      keyIdentifier = getKeyIdentifier(options);
      progress = cli.interaction.progress(util.format($("Requesting a backup blob for key %s"), keyIdentifier)); return (function ___(__then) { (function ___(_) { __tryCatch(_, function __$__8() {

            return client.backupKey(options.vaultUri, options.keyName, __cb(_, __frame, 40, 22, function ___(__0, __1) { blob = __1; _(null, null, true); }, true)); }); })(function ___(__e, __r, __cont) { (function ___(__then) { __tryCatch(_, function __$__8() {


              progress.end(); __then(); }); })(function ___() { __tryCatch(_, function ___() { if (__cont) { __then(); } else { _(__e, __r); }; }); }); }); })(function ___() { __tryCatch(_, function __$__8() {


          log.info(util.format($("Writing file %s"), options.outputFile));
          fs.writeFileSync(options.outputFile, blob.value); _(); }); }); }); });


  key.command("restore [vault-name] [input-file]").description($("Restores a backed up key to a vault")).usage("[options] <vault-name> <input-file>").option("-u, --vault-name <vault-name>", $("the vault name")).option("-f, --input-file <input-file>", $("name of the binary file that contains backup data")).execute(function __9(vaultName, inputFile, options, _) { var buffer, client, key, progress; var __frame = { name: "__9", line: 601 }; return __func(_, this, arguments, __9, 3, __frame, function __$__9() {










      log.verbose(("arguments: " + JSON.stringify({
        vaultName: vaultName,
        inputFile: inputFile,
        options: options })));


      options.vaultName = (options.vaultName || vaultName);
      options.inputFile = (options.inputFile || inputFile);

      if (!options.vaultName) {
        return _(null, cli.missingArgument("vault-name")); } ;


      if (!options.inputFile) {
        return _(null, cli.missingArgument("input-file")); } ;






      log.info(util.format($("Reading file %s"), options.inputFile));
      buffer = fs.readFileSync(options.inputFile);

      options.vaultUri = createVaultUri(options);
      client = createClient(options);


      progress = cli.interaction.progress(util.format($("Restoring key into vault %s"), options.vaultUri)); return (function ___(__then) { (function ___(_) { __tryCatch(_, function __$__9() {

            return client.restoreKey(options.vaultUri, buffer, __cb(_, __frame, 36, 21, function ___(__0, __1) { key = __1; _(null, null, true); }, true)); }); })(function ___(__e, __r, __cont) { (function ___(__then) { __tryCatch(_, function __$__9() {

              progress.end(); __then(); }); })(function ___() { __tryCatch(_, function ___() { if (__cont) { __then(); } else { _(__e, __r); }; }); }); }); })(function ___() { __tryCatch(_, function __$__9() {


          showKey(key); _(); }); }); }); });



  function createVaultUri(options) {
    var subscription = profile.current.getSubscription(options.subscription);
    return (("https://" + options.vaultName) + subscription.keyVaultDnsSuffix); };


  function createClient(options) {
    var subscription = profile.current.getSubscription(options.subscription);
    log.verbose(util.format($("Using subscription %s (%s)"), subscription.name, subscription.id));
    return utils.createKeyVaultClient(subscription); };


  function isPemEncrypted(pem) {
    var msg = forge.pem.decode(pem)[0];
    if ((!msg || ((((msg.type !== "ENCRYPTED PRIVATE KEY") && (msg.type !== "PRIVATE KEY")) && (msg.type !== "RSA PRIVATE KEY"))))) {
      return false; } ;

    return ((msg.procType && (msg.procType.type === "ENCRYPTED"))); };


  function getKeyIdentifier(options) {
    var kid = ((options.vaultUri + "/keys/") + options.keyName);
    if (options.keyVersion) {
      kid += ("/" + options.keyVersion); } ;

    return kid; };


  function parseKeyPropertiesArguments(vaultName, keyName, keyVersion, options, requireDestination) {

    log.verbose(("arguments: " + JSON.stringify({
      vaultName: vaultName,
      keyName: keyName,
      keyVersion: keyVersion,
      options: options })));


    options.vaultName = (options.vaultName || vaultName);
    options.keyName = (options.keyName || keyName);
    options.keyVersion = (options.keyVersion || keyVersion);

    if (!options.vaultName) {
      return cli.missingArgument("vault-name"); } ;


    if (!options.keyName) {
      return cli.missingArgument("key-name"); } ;


    if ((requireDestination && !options.destination)) {
      return cli.missingArgument("destination"); } ;


    if (options.destination) {
      var kty;
      __.each(KEY_DEST_TYPE_MAP, function(value, key) {
        if (utils.ignoreCaseEquals(key, options.destination)) {
          kty = value; } ; });



      if (!kty) {
        throw new Error(util.format($("Invalid value for destination argument. Accepted values are: %s"), KEY_DESTS.join(", "))); } ;


      options.hsm = ((kty === "RSA-HSM")); }
     else {
      options.hsm = false; } ;


    options.keyOps = kvUtils.parseArrayArgument("key-ops", options.keyOps, KEY_OPS, []);
    options.enabled = kvUtils.parseBooleanArgument("enabled", options.enabled, true);
    options.tags = kvUtils.parseTagsArgument("tags", options.tags); };



  function getKeyVersions(client, vaultUri, keyName, _) { var keys, result; var __frame = { name: "getKeyVersions", line: 721 }; return __func(_, this, arguments, getKeyVersions, 3, __frame, function __$getKeyVersions() {

      log.verbose(util.format($("Loading versions of key %s"), keyName));

      keys = [];
      return client.getKeyVersions(vaultUri, keyName, null, __cb(_, __frame, 5, 24, function ___(__0, __1) { result = __1; return (function __$getKeyVersions(__then) {
          if (result) {
            keys = result; return (function ___(__break) { var __more; var __loop = __cb(_, __frame, 0, 0, function __$getKeyVersions() { __more = false;

                var __4 = (result && result.nextLink); if (__4) {
                  log.verbose(util.format($("Found %d versions, loading more"), keys.length));
                  return client.getKeyVersionsNext(result.nextLink, __cb(_, __frame, 11, 24, function ___(__0, __2) { result = __2;
                    if (result) {
                      keys = keys.concat(result); } ; while (__more) { __loop(); }; __more = true; }, true)); } else { __break(); } ; }); do { __loop(); } while (__more); __more = true; })(__then); } else { __then(); } ; })(function __$getKeyVersions() {



          return _(null, keys); }); }, true)); }); };


  function showKey(key) {
    key.key.n = kvUtils.bufferToBase64Url(key.key.n);
    key.key.e = kvUtils.bufferToBase64Url(key.key.e);

    cli.interaction.formatOutput(key, function(key) {
      key.attributes = kvUtils.getAttributesWithPrettyDates(key.attributes);
      utils.logLineFormat(key, log.data); }); };



  function showKeyRow(row, item) {
    var identifier = kvUtils.parseKeyIdentifier(item.kid);


    row.cell($("Name"), identifier.name);
    if (identifier.version) {
      row.cell($("Version"), identifier.version); } ;

    row.cell($("Enabled"), item.attributes.enabled);
    var attributes = kvUtils.getAttributesWithPrettyDates(item.attributes);
    row.cell($("Not Before"), (attributes.notBefore || ""));
    row.cell($("Expires"), (attributes.expires || ""));
    row.cell($("Created"), attributes.created);
    row.cell($("Updated"), attributes.updated); };


  function setRsaParameters(dest, key) {
    dest.n = bigIntegerToBuffer(key.n);
    dest.e = bigIntegerToBuffer(key.e);
    dest.d = bigIntegerToBuffer(key.d);
    dest.p = bigIntegerToBuffer(key.p);
    dest.q = bigIntegerToBuffer(key.q);
    dest.dp = bigIntegerToBuffer(key.dP);
    dest.dq = bigIntegerToBuffer(key.dQ);
    dest.qi = bigIntegerToBuffer(key.qInv); };


  function bigIntegerToBuffer(n) {

    var data = n.toByteArray();
    var leadingZeroes = 0;
    while (((leadingZeroes < data.length) && (data[leadingZeroes] === 0))) {
      ++leadingZeroes; };

    if (leadingZeroes) {
      data = data.slice(leadingZeroes); } ;


    return new Buffer(data); };};
