// azure-cli/lib/commands/arm/network/networkUsage.js
/*** Generated by streamline 0.10.17 (callbacks) - DO NOT EDIT ***/ var __rt=require('streamline/lib/callbacks/runtime').runtime(__filename, false),__func=__rt.__func,__cb=__rt.__cb; var __ = require("underscore");















var utils = require("../../../util/utils");
var $ = utils.getLocaleString;

function NetworkUsage(cli, networkManagementClient) {
  this.networkManagementClient = networkManagementClient;
  this.output = cli.output;
  this.interaction = cli.interaction;};


__.extend(NetworkUsage.prototype, {




  list: function list__1(location, options, _) { var self, output, usages, __this = this; var __frame = { name: "list__1", line: 31 }; return __func(_, this, arguments, list__1, 2, __frame, function __$list__1() { self = __this;

      output = self.output;
      return self.networkManagementClient.usages.list(location, __cb(_, __frame, 3, 50, function ___(__0, __1) { usages = __1;

        self.interaction.formatOutput(usages, function(outputData) {
          output.table(outputData, function(row, item) {
            row.cell($("Name"), item.name.localizedValue);
            row.cell($("Unit"), item.unit);
            row.cell($("CurrentValue"), item.currentValue);
            row.cell($("Limit"), item.limit); }); }); _(); }, true)); }); }});










module.exports = NetworkUsage;
